components/timezone/src/ixdtf.rs

```diff
@@ -0,0 +1,424 @@
+        Self::Syntax(value)
+    }
+}
+
+impl From<RangeError> for ParseError {
+    fn from(value: RangeError) -> Self {
+        Self::Range(value)
+    }
+}
+
+impl From<InvalidOffsetError> for ParseError {
+    fn from(_: InvalidOffsetError) -> Self {
+        Self::InvalidOffsetError
+    }
+}
+
+impl TryFrom<UTCOffsetRecord> for GmtOffset {
+    type Error = InvalidOffsetError;
+
+    fn try_from(value: UTCOffsetRecord) -> Result<Self, Self::Error> {
+        let hour_seconds = i32::from(value.hour) * 3600;
+        let minute_seconds = i32::from(value.minute) * 60;
+        Self::try_from_offset_seconds(
+            i32::from(value.sign as i8) * (hour_seconds + minute_seconds + i32::from(value.second)),
+        )
+    }
+}
+
+// ==== CustomTimeZone methods and traits ====
+
+impl CustomTimeZone {
+    /// Create a [`CustomTimeZone`] from an IXDTF syntax string.
+    ///
+    /// ✨ *Enabled with the `ixdtf` Cargo feature.*
+    ///
+    /// # Examples
+    ///
+    /// An IXDTF string can provide a time zone in two parts: the DateTime UTC Offset or the Time Zone
+    /// Annotation. A DateTime UTC Offset is the time offset as laid out by RFC3339; meanwhile, the Time
+    /// Zone Annotation is the annotation laid out by RFC9557 and is defined as a UTC offset or IANA Time
+    /// Zone identifier.
+    ///
+    /// ## DateTime UTC Offsets
+    ///
+    /// Below is an example of a time zone from a DateTime UTC Offset. The syntax here is familiar to a RFC3339
+    /// DateTime string.
+    ///
+    /// ```
+    /// use icu_timezone::{CustomTimeZone, GmtOffset};
+    ///
+    /// let tz_from_offset = CustomTimeZone::try_from_ixdtf_str("2024-08-08T12:08:19-05:00").unwrap();
+    ///
+    /// assert_eq!(tz_from_offset, CustomTimeZone {
+    ///     gmt_offset: Some(GmtOffset::try_from_offset_seconds(-18000).unwrap()),
+    ///     time_zone_id: None,
+    ///     metazone_id: None,
+    ///     zone_variant: None,
+    /// });
+    ///
+    /// ```
+    ///
+    /// ## Time Zone Annotations
+    ///
+    /// Below is an example of a time zone being provided by a time zone annotation.
+    ///
+    /// ```
+    /// use icu_timezone::{CustomTimeZone, GmtOffset, TimeZoneBcp47Id};
+    /// use tinystr::tinystr;
+    ///
+    /// let tz_from_offset_annotation = CustomTimeZone::try_from_ixdtf_str("2024-08-08T12:08:19[-05:00]").unwrap();
+    /// let tz_from_iana_annotation = CustomTimeZone::try_from_ixdtf_str("2024-08-08T12:08:19[America/Chicago]").unwrap();
+    ///
+    /// assert_eq!(tz_from_offset_annotation, CustomTimeZone {
+    ///     gmt_offset: Some(GmtOffset::try_from_offset_seconds(-18000).unwrap()),
+    ///     time_zone_id: None,
+    ///     metazone_id: None,
+    ///     zone_variant: None,
+    /// });
+    ///
+    /// assert_eq!(tz_from_iana_annotation, CustomTimeZone {
+    ///     gmt_offset: None,
+    ///     time_zone_id: Some(TimeZoneBcp47Id(tinystr!(8, "uschi"))),
+    ///     metazone_id: None,
```

Comment:
Issue: I think you should fill in the `metazone_id`, which is computable from a `time_zone_id` and a local datetime, both of which which you have available in these IXDTF strings. See `MetazoneCalculator`.

Probably fine to leave it empty if you don't have a full datetime in the IXDTF string (e.g. if the time of day is missing).